daemon/containerd/image_push.go

```diff
@@ -2,14 +2,256 @@ package containerd
+			return errdefs.NotImplemented(errors.New("push all tags is not implemented"))
+		}
+	}
+
+	leasedCtx, release, err := i.client.WithLease(ctx)
+	if err != nil {
+		return err
+	}
+	defer release(leasedCtx)
+	out := streamformatter.NewJSONProgressOutput(outStream, false)
+
+	img, err := i.client.ImageService().Get(ctx, targetRef.String())
+	if err != nil {
+		return errdefs.NotFound(err)
+	}
+
+	target := img.Target
+	store := i.client.ContentStore()
+
+	resolver, tracker := i.newResolverFromAuthConfig(authConfig)
+	pushProgress := pushProgress{Tracker: tracker}
+	jobs := newJobs()
+	finishProgress := jobs.showProgress(ctx, out, combinedProgress([]progressUpdater{
+		&pushProgress,
+		pullProgress{ShowExists: false, Store: store},
+	}))
+	defer finishProgress()
+
+	var limiter *semaphore.Weighted = nil // TODO: Respect max concurrent downloads/uploads
+
+	mountableBlobs, err := i.findMissingMountable(ctx, store, jobs, target, targetRef, limiter)
+	if err != nil {
+		return err
+	}
+	for dgst := range mountableBlobs {
+		pushProgress.addMountable(dgst)
+	}
+
+	// Create a store which fakes the local existence of possibly mountable blobs.
+	// Otherwise they can't be pushed at all.
+	realStore := store
+	wrapped := wrapWithFakeMountableBlobs(store, mountableBlobs)
+	store = wrapped
+
+	pusher, err := resolver.Pusher(ctx, targetRef.String())
+	if err != nil {
+		return err
+	}
+
+	addChildrenToJobs := containerdimages.HandlerFunc(
+		func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+			children, err := containerdimages.Children(ctx, store, desc)
+			if err != nil {
+				return nil, err
+			}
+			for _, c := range children {
+				jobs.Add(c)
+			}
+
+			jobs.Add(desc)
+
+			return nil, nil
+		},
+	)
+
+	appendSource, err := docker.AppendDistributionSourceLabel(realStore, targetRef.String())
+	if err != nil {
+		// This shouldn't happen at this point because the reference would have to be invalid
+		// and if it was, then it would error out earlier.
+		return errdefs.Unknown(errors.Wrap(err, "failed to create an handler that appends distribution source label to pushed content"))
+	}
+
+	handlerWrapper := func(h images.Handler) images.Handler {
+		return containerdimages.Handlers(addChildrenToJobs, h, appendSource)
+	}
+
+	err = remotes.PushContent(ctx, pusher, target, store, limiter, platforms.All, handlerWrapper)
+	if err != nil {
+		if containerdimages.IsIndexType(target.MediaType) {
+			if cerrdefs.IsNotFound(err) {
+				err = errdefs.NotFound(fmt.Errorf(
+					"missing content: %w\n"+
+						"Note: You're trying to push a manifest list/index which "+
+						"references multiple platform specific manifests, but not all of them are available locally "+
+						"or available to the remote repository.\n"+
+						"Make sure you have all the referenced content and try again.",
+					err))
+			}
+		}
+	}
+	return err
+}
+
+// prepareMissing will walk the target descriptor recursively and returns up
```

Comment:
Looks like the docs doesn't match the function name

Reply:
Thanks! Fixed.